Interop/Runtime/src/main/kotlin/kotlinx/cinterop/Utils.kt

```diff
@@ -69,30 +69,61 @@ public abstract class AutofreeScope : DeferScope(), NativePlacement {
     abstract override fun alloc(size: Long, align: Int): NativePointed
 }
 
+private const val ChunkSize = 4096
+
 public open class ArenaBase(private val parent: NativeFreeablePlacement = nativeHeap) : AutofreeScope() {
 
-    private var lastChunk: NativePointed? = null
+    private var lastLargeObjectsChunk: NativePointed? = null
+    private var lastSmallObjectsChunk: NativePointed? = null
+    private var offset = 0
 
-    final override fun alloc(size: Long, align: Int): NativePointed {
+    private fun allocLargeObject(size: Long, align: Int): NativePointed {
         // Reserve space for a pointer:
         val gapForPointer = maxOf(pointerSize, align)
 
         val chunk = parent.alloc(size = gapForPointer + size, align = gapForPointer)
-        nativeMemUtils.putNativePtr(chunk, lastChunk.rawPtr)
-        lastChunk = chunk
+        nativeMemUtils.putNativePtr(chunk, lastLargeObjectsChunk.rawPtr)
+        lastLargeObjectsChunk = chunk
         return interpretOpaquePointed(chunk.rawPtr + gapForPointer.toLong())
     }
 
+    final override fun alloc(size: Long, align: Int): NativePointed {
+        if (size >= ChunkSize / 2)
+            return allocLargeObject(size, align)
+
+        val startPtr = lastSmallObjectsChunk?.rawPtr?.toLong()
+        if (startPtr != null) {
+            val ptr = (startPtr + offset + align - 1) and (align - 1).toLong().inv()
+            if (ptr + size <= startPtr + ChunkSize.toLong())
+            {
+                offset = (ptr + size - startPtr).toInt()
+                return interpretOpaquePointed(nativeNullPtr + ptr)
+            }
+        }
+
+        val chunk = parent.alloc(ChunkSize, align = pointerSize)
```

Comment:
I expect that typical use case for `ArenaBase` is `memScoped`, which is typically used to allocate one or two small objects.
Which case is optimized here then?

Reply:
That's an interesting question, I saw a small improvement in compilation speed after this fix, I'll try to get more info about it.

Reply:
I finally managed to reproduce the improvement (it is small, about 0.5% overall), and the actual case is the aforementioned 2 small objects allocation: in the current implementation there are two calls to `nativeMemUtils.alloc`, while in mine - only one (with a trade-off in increased memory footprint, the actual chunk size may be discussed).

Reply:
Can the same improvement be achieved by eliminating redundant `interpretOpaquePointed` calls (and replacing the remaining ones with direct constructor calls)?